primitives/state-machine/src/proving_backend.rs

```diff
@@ -109,9 +109,55 @@ impl<'a, S, H> ProvingBackendRecorder<'a, S, H>
 	}
 }
 
-/// Global proof recorder, act as a layer over a hash db for recording queried
-/// data.
-pub type ProofRecorder<H> = Arc<RwLock<HashMap<<H as Hasher>::Out, Option<DBValue>>>>;
+/// Global proof recorder, act as a layer over a hash db for recording queried data.
+#[derive(Clone, Default)]
+pub struct ProofRecorder<Hash> {
+	/// All the records that we have stored so far.
+	records: Arc<RwLock<HashMap<Hash, Option<DBValue>>>>,
+	/// The encoded size of all recorded values.
+	encoded_size: Arc<AtomicUsize>,
+}
+
+impl<Hash: std::hash::Hash + Eq> ProofRecorder<Hash> {
+	/// Record the given `key` => `val` combination.
+	pub fn record(&self, key: Hash, val: Option<DBValue>) {
+		if let Some(ref val) = val {
+			self.encoded_size.fetch_add(val.encoded_size(), Ordering::Relaxed);
+		}
+
+		self.records.write().insert(key, val);
+	}
+
+	/// Returns the value at the given `key`.
+	pub fn get(&self, key: &Hash) -> Option<Option<DBValue>> {
+		self.records.read().get(key).cloned()
+	}
+
+	/// Returns the estimated encoded size of the proof.
+	///
+	/// The estimation is maybe bigger (by in maximum 4 bytes), but never smaller than the actual
+	/// encoded proof.
+	pub fn estimate_encoded_size(&self) -> usize {
+		self.encoded_size.load(Ordering::Relaxed)
```

Comment:
why `Ordering::Relaxed` everywhere?

Reply:
Because we don't need any other ordering?

Reply:
I just don't get why everything in substrate is using relaxed ordering.
You can also put encoded_size under `Arch<RwLock<>>` along with HashMap, so that it doesn't need to be an atomic that goes out of sync if you call `record` and then `estimate_encoded_size`

Reply:
> I just don't get why everything in substrate is using relaxed ordering.

If there is no requirement that an atomic needs to be executed in some special order, we can just `Relaxed`. Here we clearly don't need any special ordering, because nothing is dependent on being updated before some mutex is locked/unlocked.

> You can also put encoded_size under Arch<RwLock<>> along with HashMap, so that it doesn't need to be an atomic that goes out of sync if you call record and then estimate_encoded_size

I know, but as you see here, I decided against this. I first thought of creating an `ProofRecorderInner` that I would put behind the mutex, but I don't wanted to create another struct etc. So, I decided to use an `Atomic`. 

The whole proof recorder could also just be some `Rc`, as we always execute block production single threaded. This stuff here is only Send/Sync for futures.

Reply:
I removed the atomic.